### DISCUSSION TITLE
Add support for CreateProjection().ForMember(string name, ...)

### DISCUSSION THREAD

**Discussion by Mithras** [LuckyPennySoftware/AutoMapper#3929] [Ideas]
Please provide an overload for [IProjectionExpression.ForMember()](https://github.com/AutoMapper/AutoMapper/blob/master/src/AutoMapper/Configuration/IMappingExpression.cs#L135) with `string name` instead of `Expression<Func<TDestination, TMember>> destinationMember` like you do in [IMappingExpression.ForMember()](https://github.com/AutoMapper/AutoMapper/blob/master/src/AutoMapper/Configuration/IMappingExpression.cs#L68).

More context. I've tried to migrate from `CreateMap()` to `CreateProjection()` but where is no drop off replacement for `ForMember(string name, ...)`. In our project we need to do something like
```
        foreach (var p in typeof(TDestination).GetProperties(BindingFlags.Public | BindingFlags.Instance))
        {
            source.ForMember(p.Name, x => x.ExplicitExpansion());
        }
```
because we use GraphQL and only map properties that are asked in a query.
In order to use `CreateProjection()` we now have to do something like this
```
        var destinationArg = Expression.Parameter(typeof(TDestination), "x");
        foreach (var p in typeof(TDestination).GetProperties(BindingFlags.Public | BindingFlags.Instance))
        {
            var propertyExpression = Expression.Property(destinationArg, p);
            var destinationMemberLambda = Expression.Lambda(propertyExpression, destinationArg);

            var configurationType = typeof(IProjectionMemberConfiguration<,,>).MakeGenericType(typeof(TSource), typeof(TDestination), p.PropertyType);
            var explicitExpansionMethod = configurationType.GetMethod("ExplicitExpansion");
            var configurationArg = Expression.Parameter(configurationType, "x");
            var explicitExpansionExpression = Expression.Call(configurationArg, explicitExpansionMethod);
            var explicitExpansionLambda = Expression.Lambda(explicitExpansionExpression, configurationArg);

            var forMemberMethod = typeof(IProjectionExpression<TSource, TDestination>).GetMethod("ForMember").MakeGenericMethod(p.PropertyType);
            forMemberMethod.Invoke(source, new object[] { destinationMemberLambda, explicitExpansionLambda.Compile() });
        }
```
which works fine but looks horrible and probably slows down our application start.

**Comment by lbargaoanu** [comment]
Try using `ForAllMaps` with `ForAllMembers` instead. Or `ForAllPropertyMaps`.

**Comment by jbogard** [comment]
We'd accept a PR for this feature request.

**Comment by lbargaoanu** [comment]
I'm not sure it's a good idea. Clearly `ForAllMembers` fits better here. And how many users we would have for such methods? People who need this kind of stuff can write an [extension method](https://github.com/AutoMapper/AutoMapper/pull/3920/commits/659eeac921a9eadcfaf702b0fd0b10a178dcfac1) or just just use `CreateMap`. I think these extra methods would only pollute intellisense and make for a worse user experience for most people.

**Reply by jbogard** [reply]
I think the original request is about a drop-in replacement for `ForMember(string, memberOptions)`, which we don't have today. But yes I think in this case a `ForAllMembers` is better solution and not just an overload that takes a string.

**Reply by Mithras** [reply]
Yes, I would totally use `ForAllMembers` but I don't think it exists for `IProjectionExpression` today.

**Comment by Mithras** [comment]
There is no `ForAllMembers` for `IProjectionExpression` as far as I can tell.
I'll create a PR if you agree it's something you want.

**Reply by lbargaoanu** [reply]
That's irrelevant, if you use `ForAllMaps` or `ForAllPropertyMaps`.

**Reply by Mithras** [reply]
It really looks like a hack to me. I just want to use property name instead of expression. Why do I have to use `ForAllMaps`/`ForAllPropertyMaps` for this?

**Reply by lbargaoanu** [reply]
So you have only one map that uses explicit expansion for all members? :)

**Reply by Mithras** [reply]
Not one but not all of them either. Only Profiles with maps related to GraphQL

**Reply by lbargaoanu** [reply]
Yes, and using `ForAllPropertyMaps` in that profile only applies there, not globally. So one call would cover everything.

**Reply by Mithras** [reply]
Got it. I'll try it later today. Thank you!

**Comment by Mithras** [comment]
Update on this. 
```
this.Internal().ForAllPropertyMaps(_ => true, (_, configuration) => configuration.ExplicitExpansion());
```
inside a Profile works fine but I still believe `IProjectionExpression.ForMember()` should have an overload which accepts property name. In my specific case I was able to workaround it by using `Internal().ForAllPropertyMaps()` but it might not be the same for other valid scenarios. I'm not even sure what's the reason behind forcing people to use expressions. In real world people have unit tests with `ConfigurationProvider.AssertConfigurationIsValid()` so I don't understand why we can't just use property name if we want to.

**Reply by lbargaoanu** [reply]
You can, just use `CreateMap` or the extension method above. `ProjectTo` is a query and queries work with types. 99.9% procent of the people don't use `ForAllMembers` or strings with `ProjectTo`, they just use dedicated DTOs. We should write _new_ APIs for them, for the main use case, not for these weird corner cases. These kind of "features" represent API noise for most people in my view.
That being said, it seems to me that you replaced some reflection code and a setting per map with one line of code. I call that an improvement :)

**Reply by Mithras** [reply]
I don't share "everything but happy path is a corner case" idea.
